src/devices/machine/acorn_memc.h

```diff
@@ -0,0 +1,102 @@
+{
+public:
+	acorn_memc_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	template <typename T>
+	acorn_memc_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock, T &&vidc_tag)
+		: acorn_memc_device(mconfig, tag, owner, clock)
+	{
+		m_vidc.set_tag(std::forward<T>(vidc_tag));
+	}
+
+	auto abort_w()   { return m_abort_w.bind(); }
+	auto sirq_w()    { return m_sirq_w.bind();  }
+
+	// enable/disable the output of the correct DRAM row and column for DRAM access.
+	// This allows to emulate the correct DRAM mirrors that are used by RISC OS to detect the installed RAM,
+	// but requires a bitswap on every access and is not required by machines with 2 or more MB of RAM.
+	void output_dram_rowcol(bool v)  { m_output_dram_rowcol = v; }
+
+	uint32_t logical_r(offs_t offset, uint32_t mem_mask = ~0);
+	void logical_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
+	void page_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
+	void registers_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
+	uint32_t high_mem_r(offs_t offset, uint32_t mem_mask = ~0);
+	void high_mem_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
+
+	DECLARE_WRITE_LINE_MEMBER(spvmd_w);
+	DECLARE_WRITE_LINE_MEMBER(sndrq_w);
+	DECLARE_WRITE_LINE_MEMBER(vidrq_w);
+
+protected:
+	// device-level overrides
+	virtual void device_resolve_objects() override;
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+	// device_memory_interface overrides
+	virtual space_config_vector memory_space_config() const override;
+
+private:
+	void memc_map_debug_commands(int ref, const std::vector<std::string> &params);
+	uint32_t dram_address(uint32_t address);
+	bool is_valid_access(int page, bool write);
+	uint32_t invalid_access(bool is_write, offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
+	void do_sound_dma();
+	void do_video_dma();
+
+	static constexpr const int m_page_sizes[4] = { 4096, 8192, 16384, 32768 };
+	optional_device<acorn_vidc10_device> m_vidc;
+	const address_space_config m_space_config;
+
+	address_space *     m_space;
+	devcb_write_line    m_abort_w;
+	devcb_write_line    m_sirq_w;
+	bool                m_output_dram_rowcol;
+	int                 m_spvmd;
+	uint8_t             m_pagesize;
+	bool                m_latchrom;
+	bool                m_video_dma_on;
+	bool                m_sound_dma_on;
+	bool                m_cursor_enabled;
+	bool                m_os_mode;
+	uint32_t            m_vidinit;
+	uint32_t            m_vidstart;
+	uint32_t            m_vidend;
+	uint32_t            m_vidcur;
+	uint32_t            m_cinit;
+	uint32_t            m_sndstart;
+	uint32_t            m_sndend;
+	uint32_t            m_sndcur;
+	uint32_t            m_sndendcur;
+	int16_t             m_pages[0x2000];        // the logical RAM area is 32 megs, and the smallest page size is 4k
+	uint8_t             m_pages_ppl[0x2000];
```

Comment:
Can it really keep state internally for enough pages to cover the entire logical space, or does it have some kind of cache and need to go out to main memory on a miss?

Reply:
The real hardware works the other way around, it has a 128 entry lockup table to map each physical page to a logical page (this is described on page 27 of the MEMC datasheet).
The old implementation (in archimds.cpp) was done in this way (I think for performance reasons) and I assumed it was correct, but now I see the problem that multiple logical pages can be mapped to the same physical page and this is impossible in the real MEMC. Anyway it should be enough to unmap duplicate pages before adding a new entry to solve this problem.

Reply:
> The old implementation (in archimds.cpp) was done in this way (I think for performance reasons) 

Been a while but I'm positive it was done to rather made it to work out first, given that RiscOS POST stress test that out and laconically locking up with no message if MEMC fails (unless some serial debugging needs to be enabled somehow).
Latest version of RiscOS available in `aa310.cpp` also always sported a non-fatal red screen on POST, is it been fixed with this PR?

Reply:
RISC OS 3 still has the same problem.

According to the document you linked RISC OS 3 in MAME fails 4 tests:

| Mask     |   Result   | Meaning                             |
| -------- | ---------- | ----------------------------------- |
| 00040000 |    Pass    | Long RAM test failure               |
| 00020000 |    Pass    | RAM control line failure            |
| 00010000 |    Pass    | CMOS unreadable                     |
| 00008000 |    Pass    | VIDC Sirq timing failed             |
| 00004000 |    Pass    | VIDC Virq timing failed             |
| 00001000 |  **Fail**  | IOC register test failed            |
| 00000800 |  **Fail**  | MEMC protection failed              |
| 00000400 |  **Fail**  | MEMC CAM mapping failed             |
| 00000200 |    Pass    | ROM failed checksum test            |
| 00000100 |  **Fail**  | CMOS RAM checksum error             |
| 00000080 |    Off     | VRAM detected                       |
| 00000040 |    Off     | PC-style IO world detected          |
| 00000020 |    Off     | Long memory test disabled           |
| 00000010 |    Off     | ARM 3 fitted                        |
| 00000008 |    On      | Long memory test performed          |
| 00000004 |    Off     | Self-test due to test link          |
| 00000002 |    Off     | Self-test due to interface hardware |
| 00000001 |    On      | Self-test due to power on           |


I investigated the 2 MEMC failures and they are both related to the missing ABORT emulation in the ARM core
- 00000400 expects a data abort exception when reading an unmapped page.
- 00000800 expects a data abort exception when writing on a page without the required privileges.

Tonight I will check the reasons for the other 2 failures.